File: HW/HW10/HW10.java

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
// function provided on canvas page
 public class TreeNode {

     int val;

     TreeNode left;

     TreeNode right;

     TreeNode() {}

     TreeNode(int val) { this.val = val; }

     TreeNode(int val, TreeNode left, TreeNode right) {

         this.val = val;

         this.left = left;

         this.right = right;

     }

  }
// end of function provided
// solution:
class HW10 {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelValues = new ArrayList<>();

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                levelValues.add(node.val);

                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            result.add(levelValues);
        }

        return result;
    }
}

public class Main {
    public static void main(String[] args) {
      // test case borrowed from canvas page
        TreeNode root = new TreeNode(4); 
        root.left = new TreeNode(3);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(1);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);

        HW10 solution = new HW10();
        List<List<Integer>> result = solution.levelOrder(root);
        System.out.println(result);
    }
}

File: HW/HW10/HW10.py
import collections
from collections import deque
from typing import List, Optional
class TreeNode(object): #node implementation

     def __init__(self, val=0, left=None, right=None):

         self.val = val

         self.left = left

         self.right = right

class HW10: #solution
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        result = [] 
        queue = deque([root])
        while queue:
            qLength = len(queue)
            level = []
            for i in range (qLength):
                node = queue.popleft()
                if node:
                    level.append(node.val)
                    queue.append(node.left)
                    queue.append(node.right)
            if level:
                result.append(level)
        return result

# test case code        
root = TreeNode(4)
root.left = TreeNode(3)
root.right = TreeNode(8)
root.left.left = TreeNode(1)
root.right.left = TreeNode(5)
root.right.right = TreeNode(9)


solution = HW10()
result = solution.levelOrder(root)
print(result) 
